include/mesh.tcc

```diff
@@ -298,84 +299,47 @@ void mpm::Mesh<Tdim>::compute_cell_neighbours() {
-                    },
-                    tbb::simple_partitioner());
+  tbb::parallel_for(
+      tbb::blocked_range<int>(size_t(0), size_t(cells_.size()),
+                              tbb_grain_size_),
+      [&](const tbb::blocked_range<int>& range) {
+        for (int i = range.begin(); i != range.end(); ++i) {
+          // Iterate over each node in current cell
+          for (auto id : cells_[i]->nodes_id()) {
+            auto cell_id = cells_[i]->id();
+            // Get the cells associated with each node
+            for (auto neighbour_id : node_cell_map[id])
+              if (neighbour_id != cell_id)
+                cells_[i]->add_neighbour(neighbour_id);
+          }
+        }
+      },
+      tbb::simple_partitioner());
 }
 
 //! Find particle neighbours for all particle
 template <unsigned Tdim>
-void mpm::Mesh<Tdim>::compute_particle_neighbours(const Cell<Tdim>& cell) {
-  // Check argument passed
-  Container<Cell<Tdim>> cell_ctr;
-  cell_ctr.add(cell);
-  auto& cells = (cell_ctr == nullptr) ? cells_ : cell_ctr;
-
-  // Check whether cells have been assigned with neighbours
-  if ((*cells_.cbegin())->nneighbours() == 0)
-    throw std::runtime_error(
-        "No neighbours have been assigned to cell, cannot "
-        "compute_particle_neighbours for single cell");
-
-  // Loop over cells
+void mpm::Mesh<Tdim>::find_particle_neighbours(const Cell<Tdim>& cell) {
+  // Find particle neighbours in all cells or a prescribed cell
+  Container<Cell<Tdim>> cell_container;
+  cell_container.add(cell);
+  auto& cells = (cell_container == nullptr) ? cells_ : cell_container;
+
   for (auto citr = cells.cbegin(); citr != cells.cend(); ++citr) {
-    // Loop over the neighbouring cell particles
+    // Particles in current cell
     std::vector<mpm::Index> neighbour_particles = (*citr)->particles();
-    for (const auto& neighbour_cell_id : (*citr)->neighbours())
-      neighbour_particles.insert(
-          neighbour_particles.end(),
-          map_cells_[neighbour_cell_id]->particles().begin(),
-          map_cells_[neighbour_cell_id]->particles().end());
-
-    // Loop over the particles in the current cells and assign particle
-    // neighbours
-    for (const auto& p_id : (*citr)->particles()) {
-      bool status =
-          map_particles_[p_id]->assign_neighbours(neighbour_particles);
-      if (!status)
-        throw std::runtime_error("Cannot assign valid particle neighbours");
+    // Loop over all neighboring cells, and append particle ids from each cell
+    for (const auto& neighbour_cell_id : (*citr)->neighbours()) {
+      // Get particle ids from each cell
+      auto& particle_ids = map_cells_[neighbour_cell_id]->particles();
+      neighbour_particles.insert(neighbour_particles.end(),
+                                 particle_ids.begin(), particle_ids.end());
     }
-  }
-}
 
-//! Find particle neighbours for a given particle
-template <unsigned Tdim>
-void mpm::Mesh<Tdim>::compute_particle_neighbours(
```

Comment:
Why this is removed? This is important for the boundary particles. Could we please revert?

Reply:
we can accomplish something similar with passing the cell to `find_particle_neighbours`, also we may have to recompute neighbors at each step, so this specialized function won't have a big effect.

Reply:
I got what you mean. I do have some points to say though. I will talk to you privately.